File: ruby/minesweeper/minesweeper.rb
class Minesweeper
  private

  attr_reader :board

  public

  def self.annotate(input)
    new(input).reveal_count
  end

  def initialize(input)
    @board = input
  end

  def reveal_count
    0.upto(board.length - 1) do |r|
      0.upto(board[r].length - 1) do |c|
        next if board[r][c] == '*'
        bombs = count_bombs(r, c)
        board[r][c] = bombs.to_s unless bombs.zero?
      end
    end
    board
  end

  private

  def count_bombs(row, col)
    count = 0
    [-1, 0, 1].each do |dr|
      [-1, 0, 1].each do |dc|
        count += 1 if (0...board.length).include?(row + dr) and
                      (0...board[row].length).include?(col + dc) and
                      board[row + dr][col + dc] == '*'
      end
    end
    count
  end
end

File: elixir/minesweeper/lib/minesweeper.ex
defmodule Minesweeper do
  @doc """
  Annotate empty spots next to mines with the number of mines next to them.
  """
  @spec annotate([String.t()]) :: [String.t()]

  def annotate(board) do
    nums = for row <- board do
      for char <- to_charlist(row) do
        if char == ?*, then: 9, else: 0
      end
    end

  end
end
